### ISSUE TITLE
Adding npm dependencies for package causing many errors

### ISSUE THREAD

**Issue by Aerlinger** [cocos-creator/editor-framework#352]
I have a generic package with the following directory structure:

```
myPackage/
  - package.json
  - main.js
  - panel/
     - panel.js
```

I wanted to use font-awesome in my package and tried running `npm install --save font-awesome`. It installed fine into a `node_modules` folder and updated the package.json.

However, after installing this module I get a long list of errors in my package:

![image](https://cloud.githubusercontent.com/assets/604138/17024994/f221e192-4f27-11e6-9cf4-1f3726ffcda0.png)

Yet, removing font-awesome from my `package.json` AND running `npm prune` resolves the issue entirely.

These errors occur for any npm packages I include in package.json, not just font-awesome.

For reference, here are the relevant files in my package:

**package.json**
```json
{
  "name": "sample-pkg",
  "version": "0.0.1",
  "description": "A placeholder panel for layout design",
  "author": "Firebox Technology",
  "main": "main.js",
  "main-menu": {
    "Examples/SamplePkg": {
      "message": "sample-pkg:open"
    }
  },
  "panel": {
    "main": "panel/panel.js",
    "popable": false,
    "type": "dockable",
    "title": "Sample Package",
    "icon": "panel/icon.png",
    "height": 300,
    "shortcuts": {
      "command+k": "clear",
      "ctrl+k": "clear"
    }
  },
  "dependencies": {
      "font-awesome": "^4.6.3"
  }
}
```

**main.js**
```javascript
'use strict';

const Electron = require('electron');

module.exports = {
  load () {
    Editor.log("Pkg loaded")
  },

  unload () {
    // Editor.Menu.unregister( 'open-log-file' );
  },

  messages: {
    'open' () {
      Editor.Panel.open('sample-pkg');
    }
  }
};
```

**panel/panel.js**
```javascript
'use strict';

Editor.Panel.extend({
  template: `
    <div id="console" class="fit">
        <header>
            <textarea id="editor"></textarea>
        </header>
    </div>
    `,

  listeners: {
    click ( event ) {
      event.stopPropagation();
      console.log('click!');
    },

    'panel-resize' ( event ) {
      console.log(event.target);
    },

    open() {
      Editor.Panel.open('sample-pkg')
    }
  },

  ready () {

  },

  run (argv) {
    console.log(`Pkg run with args: ${argv}`);
  }
});
```

Not sure if I'm missing anything, but it seems like adding an npm module without any other change to the code shouldn't cause these issues.

**Comment by Aerlinger** [comment]
Interestingly, the same issue seems to happen with Bower as well.

**Comment by VinciShark** [comment]
I think Editor-Framework just forget that panel can require its own npm requirements and bowers'

**Comment by jwu** [comment]
Hi @Aerlinger , I cannot reproduce the problem :(

I add an example package which use a npm module locally without problem, here it is: https://github.com/fireball-packages/hexdumper

**Comment by Aerlinger** [comment]
Interesting, I don't get any errors for that hexdumper example. Instead I just get a blank window when adding it to example-apps and clicking `Tools`->`Hexdumper`:

![image](https://cloud.githubusercontent.com/assets/604138/17099210/f254c3a4-5234-11e6-8511-0dc03b4baf5a.png)

Is there an example file for it?

**Comment by jwu** [comment]
try to drag a *small size* file onto hex dumper.

**Comment by Aerlinger** [comment]
The issue seems related to how `node_modules` dependencies are looked up in the `package-search-path`. Adding your hexdumper package to `example-apps/demo` works as expected.  However, I have a simple application which I've forked from `example-apps/demo` and is identical save for an additional package folder defined in `package-search-path` named `my-packages`. 

```javascript
init ( opts, cb ) {
    let settingsPath = Path.join(Editor.App.path, '.settings');

    Editor.init({
      'profile': {
        local: settingsPath,
      },
      'package-search-path': [
        Editor.url('app://packages/'),
        Editor.url('app://package-examples/'),
        // Editor.url('app://my-packages/'), Uncommenting this yields an error
      ],
      'panel-window': 'app://window.html',
      'layout': Editor.url('app://layout.json'),
    });

    if ( cb ) {
      cb ();
    }
  },
```

The error occurs when both of these conditions are true:
- `Editor.url('app://my-packages/')` is defined in `package-load-path`
- I add a `node_modules` folder to *any* package within `my-packages`

So it must be something in my addition "my-packages" folder, though the folder and package structure is identical to the demo app.

Here's the error I see from the console:

```
===== Initializing Editor =====
arguments: 
  --debug=3030
  --dev
  --show-devtools

[57351] protocol editor-framework registered
[57351] protocol app registered
[57351] protocol packages registered
[57351] protocol theme registered
[57351] Loading packages
[57351] console loaded
[57351] ipc-debugger loaded
[57351] package-manager loaded
[57351] tester loaded
[57351] ui-kit-preview loaded
[57351] demo-confirm-close loaded
[57351] demo-lifecycle loaded
[57351] demo-list-view loaded
[57351] Code editor loaded
[57351] react-example loaded
[57351] demo-run-command loaded
[57351] demo-selection loaded
[57351] demo-shortcuts loaded
[57351] demo-simple-webview loaded
[57351] demo-simple loaded
[57351] Code editor loaded
[57351] code-editor loaded
false
[57351] Dummy panel loaded
[57351] dummy-panel loaded
[57351] Code editor loaded
[57351] react-example loaded
[57351] Watching packages
[57351] Run Application
[57351] Error: ReferenceError: Pace is not defined
    at file:///Users/Aerlinger/Dropbox/Github/editor-framework/example-apps/habilis/index.html:36:7
[57351] Error: ReferenceError: Polymer is not defined
    at Object.PolymerUtils.import (/Users/Aerlinger/Dropbox/Github/editor-framework/lib/renderer/ui/utils/polymer-utils.js:101:3)
    at Object.PanelLoader.load (/Users/Aerlinger/Dropbox/Github/editor-framework/lib/renderer/panel-loader.js:121:27)
    at Ipc.sendToMain (/Users/Aerlinger/Dropbox/Github/editor-framework/lib/renderer/panel.js:43:17)
    at EventEmitter.ipcRenderer.on (/Users/Aerlinger/Dropbox/Github/editor-framework/lib/renderer/ipc.js:317:19)
    at emitMany (events.js:127:13)
    at EventEmitter.emit (events.js:201:7)
[57351] Error: ReferenceError: Polymer is not defined
    at Object.PolymerUtils.import (/Users/Aerlinger/Dropbox/Github/editor-framework/lib/renderer/ui/utils/polymer-utils.js:101:3)
    at Object.PanelLoader.load (/Users/Aerlinger/Dropbox/Github/editor-framework/lib/renderer/panel-loader.js:121:27)
    at Ipc.sendToMain (/Users/Aerlinger/Dropbox/Github/editor-framework/lib/renderer/panel.js:43:17)
    at EventEmitter.ipcRenderer.on (/Users/Aerlinger/Dropbox/Github/editor-framework/lib/renderer/ipc.js:317:19)
    at emitMany (events.js:127:13)
    at EventEmitter.emit (events.js:201:7)
[57351] Error: ReferenceError: Polymer is not defined
    at Object.PolymerUtils.import (/Users/Aerlinger/Dropbox/Github/editor-framework/lib/renderer/ui/utils/polymer-utils.js:101:3)
    at Object.PanelLoader.load (/Users/Aerlinger/Dropbox/Github/editor-framework/lib/renderer/panel-loader.js:121:27)
    at Ipc.sendToMain (/Users/Aerlinger/Dropbox/Github/editor-framework/lib/renderer/panel.js:43:17)
    at EventEmitter.ipcRenderer.on (/Users/Aerlinger/Dropbox/Github/editor-framework/lib/renderer/ipc.js:317:19)
    at emitMany (events.js:127:13)
    at EventEmitter.emit (events.js:201:7)
```


Here's the error from DevTools console:

```
theme://elements/resizer.css Failed to load resource: net::ERR_UNKNOWN_URL_SCHEME
theme://elements/tab.css Failed to load resource: net::ERR_UNKNOWN_URL_SCHEME
theme://elements/tabs.css Failed to load resource: net::ERR_UNKNOWN_URL_SCHEME
theme://elements/dock.css Failed to load resource: net::ERR_UNKNOWN_URL_SCHEME
theme://elements/panel.css Failed to load resource: net::ERR_UNKNOWN_URL_SCHEME
theme://elements/panel-frame.css Failed to load resource: net::ERR_UNKNOWN_URL_SCHEME
theme://elements/button.css Failed to load resource: net::ERR_UNKNOWN_URL_SCHEME
theme://elements/checkbox.css Failed to load resource: net::ERR_UNKNOWN_URL_SCHEME
theme://elements/color-picker.css Failed to load resource: net::ERR_UNKNOWN_URL_SCHEME
theme://elements/color.css Failed to load resource: net::ERR_UNKNOWN_URL_SCHEME
theme://elements/hint.css Failed to load resource: net::ERR_UNKNOWN_URL_SCHEME
theme://elements/input.css Failed to load resource: net::ERR_UNKNOWN_URL_SCHEME
theme://elements/loader.css Failed to load resource: net::ERR_UNKNOWN_URL_SCHEME
theme://elements/markdown.css Failed to load resource: net::ERR_UNKNOWN_URL_SCHEME
theme://elements/num-input.css Failed to load resource: net::ERR_UNKNOWN_URL_SCHEME
theme://elements/prop.css Failed to load resource: net::ERR_UNKNOWN_URL_SCHEME
theme://elements/section.css Failed to load resource: net::ERR_UNKNOWN_URL_SCHEME
theme://elements/select.css Failed to load resource: net::ERR_UNKNOWN_URL_SCHEME
theme://elements/slider.css Failed to load resource: net::ERR_UNKNOWN_URL_SCHEME
theme://elements/text-area.css Failed to load resource: net::ERR_UNKNOWN_URL_SCHEME
app://bower_components/pace/pace.js Failed to load resource: net::ERR_UNKNOWN_URL_SCHEME
app://bower_components/pace/themes/red/pace-theme-center-simple.css Failed to load resource: net::ERR_UNKNOWN_URL_SCHEME
app://bower_components/webcomponentsjs/webcomponents-lite.min.js Failed to load resource: net::ERR_UNKNOWN_URL_SCHEME
app://bower_components/polymer/polymer.html Failed to load resource: net::ERR_UNKNOWN_URL_SCHEME
app://bower_components/iron-flex-layout/iron-flex-layout.html Failed to load resource: net::ERR_UNKNOWN_URL_SCHEME
app://bower_components/fontawesome/css/font-awesome.min.css Failed to load resource: net::ERR_UNKNOWN_URL_SCHEME
app://bower_components/codemirror/lib/codemirror.css Failed to load resource: net::ERR_UNKNOWN_URL_SCHEME
app://bower_components/codemirror/lib/codemirror.js Failed to load resource: net::ERR_UNKNOWN_URL_SCHEME
index.html:36 Uncaught ReferenceError: Pace is not defined
theme://globals/fontello.css Failed to load resource: net::ERR_UNKNOWN_URL_SCHEME
3
/Users/Aerlinger/Dropbox/Github/editor-framework/lib/renderer/ui/utils/polymer-utils.js:101 Uncaught ReferenceError: Polymer is not defined
packages://ui-kit-preview/panel/button-preview.js Failed to load resource: net::ERR_UNKNOWN_URL_SCHEME
/Users/Aerlinger/Dropbox/Github/editor-framework/example-apps/habilis/packages/ui-kit-preview/panel…:157 Uncaught (in promise) TypeError: initFn is not a function(…)
﻿
```

**Comment by Aerlinger** [comment]
Another strange issue: I thought perhaps the bower dependencies weren't installed or referenced correctly so I tried `rm -Rf bower_components` and `bower cache clean` followed by `bower install` from within the directory of the problem app (called "habilis" in the above stack trace) with no success.

However, I then copied the `habilis` app folder to another folder called `habilis2` with identify file and folder structure to `habilis`. When I ran `npm run start ./example-apps/habilis2` it failed with a similar error, but in a different error order. When I repeated the `rm -Rf bower_components`, `bower cache clean`, `bower install` steps the "habilis2" app ran without issue.

**Comment by Aerlinger** [comment]
OK, after some modification I'm able to reproduce the issue with your `hexdumper` package via the following steps:

1. Clone the latest from `sample-apps`
2. Add the hexdumper package to `demo/package-examples` in sample-apps
3. Launch the demo app -- *hexdumper and everything else should run fine*
4. Now add `react: "*"` to the dependencies in hexdumper's package.json:
```
"dependencies": {
    "hexy": "^0.2.7",
    "react": "*"
  }
```
5. `npm install`
6. Re-launch the demo app.
7. You should see a stacktrace similar to the issue above.

In short, adding react, or a similar package to a package's dependencies seems to break the app. Conversely, removing `"react": *` and running `npm prune` resolves the issue, at least for me.

My thoughts are that the package's `node_modules` are somehow being loaded asynchronously, so there may be a race condition where modules are being required before they are fully loaded.

**Comment by Aerlinger** [comment]
A few other points:

- This issue comes up with packages other than react, it seems larger packages are more likely to cause the issue.
- The error occurs only if the package is in the package's `node_modules` folder. Removing the package from `"dependencies"` in package.json won't make the error go away unless `npm prune` is called or the package is otherwise removed from `node_modules`.

**Comment by jwu** [comment]
@Aerlinger thanks for your information, I can now 100% reproduce this problem. It seems like this problem is relative with https://github.com/electron/electron/issues/6258.

Let me dive deeper into it.

**Comment by jwu** [comment]
I'm trying to solve the problem by #381 , but as @Aerlinger mentioned in last commit, large packages still can broke the custom protocol.

I am now trying to write a simple electron example to reproduce the problem and submit it to Electron team.

**Comment by jwu** [comment]
Looks like this is an Electron bug, I just write a simple code and report to them: https://github.com/jwu/electron-custom-protocol-bug

**Comment by VinciShark** [comment]
Electron‘s terrible bug……

**Comment by Aerlinger** [comment]
@jwu It seems to be a race condition where the protocol is called before it's registered with Electron. Interestingly, if I modify your example to define the window outside of the watcher callback, it works:

```javascript
Electron.app.on('ready', () => {

  // register protocol app://
  Electron.protocol.registerFileProtocol('app', (request, cb) => {
    if ( !request.url ) {
      cb (-3); // ABORTED
      return;
    }

    let url = decodeURIComponent(request.url);
    let uri = Url.parse(url);

    let relativePath = uri.hostname;
    if ( uri.pathname ) {
      relativePath = Path.join( relativePath, uri.pathname );
    }

    let file = Path.join( __dirname, relativePath );
    cb ( { path: file } );
  }, err => {
    if ( err ) {
      console.log( 'Failed to register protocol app, %s', err.message );
      return;
    }
    console.log( 'protocol app registered' );
  });

  let mainWindow = new Electron.BrowserWindow({
    title: 'Main',
    width: 800,
    height: 600,
  });

  // Defined outside the chokidoar callback
  let watcher = Chokidar.watch([__dirname], {
    ignoreInitial: true,
    persistent: true,
  });

  watcher.on('ready', () => {
    // create main window and show it
    mainWindow = new Electron.BrowserWindow({
      title: 'Main',
      width: 800,
      height: 600,
    });

    mainWindow.loadURL(`file://${__dirname}/index.html`);
    mainWindow.show();
    mainWindow.openDevTools();
  });
});
```

**Comment by Aerlinger** [comment]
Looks like it was a GC issue. There's a fix on Electron master: https://github.com/electron/electron/pull/6674.

**Comment by jwu** [comment]
Cheers :)

**Comment by VinciShark** [comment]
@Aerlinger So we can just sit and wait for electron's new release?
